include/node.tcc

```diff
@@ -17,23 +17,80 @@ mpm::Node<Tdim, Tdof, Tnphases>::Node(
   // Clear any velocity constraints
   velocity_constraints_.clear();
   concentrated_force_.setZero();
+
+  // Initialize scalar properties
+  scalar_properties_.reserve(5);
+  // Mass
+  scalar_properties_.emplace_back(Eigen::Matrix<double, 1, Tnphases>::Zero());
+  // Volume
+  scalar_properties_.emplace_back(Eigen::Matrix<double, 1, Tnphases>::Zero());
+  // MassDensity
+  scalar_properties_.emplace_back(Eigen::Matrix<double, 1, Tnphases>::Zero());
+  // MassPressure
+  scalar_properties_.emplace_back(Eigen::Matrix<double, 1, Tnphases>::Zero());
+  // Pressure
+  scalar_properties_.emplace_back(Eigen::Matrix<double, 1, Tnphases>::Zero());
+
+  // Initialize vector properties
+  vector_properties_.resize(6);
+  // Displacement
+  vector_properties_.emplace_back(
+      Eigen::Matrix<double, Tdim, Tnphases>::Zero());
+  // Velocity
+  vector_properties_.emplace_back(
+      Eigen::Matrix<double, Tdim, Tnphases>::Zero());
+  // Acceleration
+  vector_properties_.emplace_back(
+      Eigen::Matrix<double, Tdim, Tnphases>::Zero());
+  // Momentum
+  vector_properties_.emplace_back(
+      Eigen::Matrix<double, Tdim, Tnphases>::Zero());
+  // ExternalForce
+  vector_properties_.emplace_back(
+      Eigen::Matrix<double, Tdim, Tnphases>::Zero());
+  // InternalForce
+  vector_properties_.emplace_back(
+      Eigen::Matrix<double, Tdim, Tnphases>::Zero());
+
   this->initialise();
 }
 
 //! Initialise nodal properties
 template <unsigned Tdim, unsigned Tdof, unsigned Tnphases>
 void mpm::Node<Tdim, Tdof, Tnphases>::initialise() noexcept {
-  mass_.setZero();
-  volume_.setZero();
-  external_force_.setZero();
-  internal_force_.setZero();
-  pressure_.setZero();
-  contact_displacement_.setZero();
-  velocity_.setZero();
-  momentum_.setZero();
-  acceleration_.setZero();
   status_ = false;
+
+  // Initialise nodal scalar properties
+  scalar_properties_.resize(5);
+  scalar_properties_.at(mpm::properties::Scalar::Mass) =
+      Eigen::Matrix<double, 1, Tnphases>::Zero();
+  scalar_properties_.at(mpm::properties::Scalar::Volume) =
+      Eigen::Matrix<double, 1, Tnphases>::Zero();
+  scalar_properties_.at(mpm::properties::Scalar::MassDensity) =
+      Eigen::Matrix<double, 1, Tnphases>::Zero();
+  scalar_properties_.at(mpm::properties::Scalar::MassPressure) =
+      Eigen::Matrix<double, 1, Tnphases>::Zero();
+  scalar_properties_.at(mpm::properties::Scalar::Pressure) =
+      Eigen::Matrix<double, 1, Tnphases>::Zero();
+
+  // Initialise nodal vector properties
+  vector_properties_.resize(6);
+  vector_properties_.at(mpm::properties::Vector::Displacement) =
+      Eigen::Matrix<double, Tdim, Tnphases>::Zero();
```

Comment:
Displacement is not needed in node, but I know why we need to append it here to avoid access error. Does that means for every variable added in mpm_properties, we need to arrange and allocate all other particles? I don't think that's efficient storage-wise.